### PR TITLE
Add browser-nativefs article

### LINKED ISSUE #2684: content: browser-nativefs [Sun May 31 2020]
* Author: @tomayac 
* Draft: TODO

### PR DESCRIPTION
**Pull request by tomayac**
Fixes #2684

Changes proposed in this pull request:

- Add the article.

### CODE REVIEW

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+    console.error(err.name, err.message);
+  }
+};
+```
+
+## Introducing browser-nativefs
+
+As perfectly fine as the Native File System API works when it is available,
+the browser support situation [currently is not great](https://caniuse.com/#feat=native-filesystem-api),
+since the API is pretty new.
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:156]:
```suggestion
As perfectly fine as the Native File System API is,
it's [not yet  widely available](https://caniuse.com/#feat=native-filesystem-api).
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    Browser support table for the Native File System API.
+    (<a href="https://caniuse.com/#feat=native-filesystem-api">Source</a>)
+  </figcaption>
+</figure>
+
+This is why I see the Native File System API as a [progressive enhancement](/progressively-enhance-your-pwa).
+As such, I want to use it when the browser supports it,
+and use the traditional approach if not;
+all while never punishing the user with unnecessary downloads of unsupported JavaScript code.
+The 💾&nbsp;[browser-nativefs](https://github.com/GoogleChromeLabs/browser-nativefs)
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:171]:
```suggestion
The [browser-nativefs](https://github.com/GoogleChromeLabs/browser-nativefs)
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+library is my answer to this challenge.
+
+### Design philosophy
+
+Since the Native File System API is still likely to change in the future,
+the browser-nativefs API is not modeled after the Native File System API,
+that is, the library is not a [polyfill](https://developer.mozilla.org/en-US/docs/Glossary/Polyfill),
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:178]:
```suggestion
the browser-nativefs API is not modeled after it.
That is, the library is not a [polyfill](https://developer.mozilla.org/en-US/docs/Glossary/Polyfill),
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+but rather a 🐴&nbsp;[ponyfill](https://github.com/sindresorhus/ponyfill).
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:179]:
```suggestion
but rather a [ponyfill](https://github.com/sindresorhus/ponyfill).
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+You can (statically or dynamically) exclusively import whatever functionality you need to keep your app as small as possible.
+The available methods are the aptly named
+[`fileOpen()`](https://github.com/GoogleChromeLabs/browser-nativefs#opening-files),
+[`directoryOpen()`](https://github.com/GoogleChromeLabs/browser-nativefs#opening-directories), and
+[`fileSave()`](https://github.com/GoogleChromeLabs/browser-nativefs#saving-files).
+Internally, the library feature-detects if the Native File System API is supported or not,
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:185]:
```suggestion
Internally, the library feature-detects if the Native File System API is supported,
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+and then imports the corresponding code path.
+
+### Using the browser-nativefs library
+
+The three methods are intuitive to use.
+You can specify the to-be-accepted `mimeTypes` or file `extensions`, and set a `multiple` flag
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:191]:
```suggestion
You can specify your app's accepted `mimeTypes` or file `extensions`, and set a `multiple` flag
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+to allow or disallow multiple files or directories to be selected.
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:192]:
```suggestion
to allow or disallow selection of multiple files or directories.
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+  // Save a file.
+  await fileSave(blob, {
+    fileName: 'Untitled.png',
+  });
+})();
+```
+
+### Demo
+
+You can see the above code in action in a 💾&nbsp;[demo](https://browser-nativefs.glitch.me/) on Glitch.
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:233]:
```suggestion
You can see the above code in action in a [demo](https://browser-nativefs.glitch.me/) on Glitch.
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+Since for security reasons cross origin sub frames are not allowed to show a file picker,
+the demo cannot be embedded in this article.
+
+## The browser-nativefs library in the wild
+
+In my free time, I contribute a tiny bit to an
+[installable PWA](https://web.dev/progressive-web-apps/#installable)
+called [Excalidraw](https://excalidraw.com/),
+a whiteboard tool that lets you easily sketch diagrams with a hand-drawn feel.
+It is fully responsive and works well from small mobile phones to big screen computers.
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:244]:
```suggestion
It is fully responsive and works well on a range of devices from small mobile phones to computers with large screens.
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+This means it needs to deal with files on all the various platforms
+that may or, more likely, may not support the Native File System API,
+which makes it a great candidate for the browser-nativefs library.
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:247]:
```suggestion
This means it needs to deal with files on all the various platforms
whether or not they support the Native File System API.
This makes it a great candidate for the browser-nativefs library.
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+
+I can, for example, start a drawing on my iPhone,
+save it (technically: download it, since Safari does not support the Native File System API)
+to my iPhone Downloads folder, open the file on my desktop (after transferring it from my phone),
+modify the file, and over-save it with my changes, or even save it as a new file.
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:252]:
```suggestion
modify the file, and over write it with my changes, or even save it as a new file.
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+<figure class="w-figure">
+  <img class="w-screenshot" src="chrome-modify.png" alt="The modified Excalidraw drawing on Chrome on the desktop.">
+  <figcaption class="w-figcaption">
+    Opening and modifying the Excalidraw drawing on the desktop where the Native File System API is supported
+    and thus the file can be accessed via the API.
+  </figcaption>
+</figure>
+
+<figure class="w-figure">
+  <img class="w-screenshot" src="chrome-oversave.png" alt="Over-saving the original file with the modifications.">
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:271]:
```suggestion
  <img class="w-screenshot" src="chrome-oversave.png" alt="Over wriing the original file with the modifications.">
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+  <figcaption class="w-figcaption">
+    Over-saving the original file with the modifications to the original Excalidraw drawing file.
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:273]:
```suggestion
    Over writing the original file with the modifications to the original Excalidraw drawing file.
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+  <img class="w-screenshot" src="chrome-save-as.png" alt="Saving the modifications to a new Excalidraw drawing file.">
+  <figcaption class="w-figcaption">
+    Saving the modifications to a new Excalidraw file. The original file remains untouched.
+  </figcaption>
+</figure>
+
+### Real life code sample
+
+Below, you can see an actual example of browser-nativefs as it is used in Excalidraw.
+The visible excerpt stems from the file
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:288]:
```suggestion
This excerpt is taken from
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+[`/src/data/json.ts`](https://github.com/excalidraw/excalidraw/blob/cd87bd6901b47430a692a06a8928b0f732d77097/src/data/jsonts#L24-L52).
+Of special interest is how the `saveAsJSON()` method passes either a file handle or `null` to browser-nativefs'
+`fileSave()` method, which causes it to over-save when a handle is given,
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:291]:
```suggestion
`fileSave()` method, which causes it to over write when a handle is given,
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+or to save to a new file if not.
+
+```js
+export const saveAsJSON = async (
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:295]: I don't see anything returned in this example either.

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+    mimeTypes: ["application/json"],
+  });
+  return loadFromBlob(blob);
+};
+```
+
+### UI considerations
+
+In the case of Excalidraw (or your app of course),
+you need to make sure the UI adopts accordingly to the browser's support situation.
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:329]:
```suggestion
Whether in Excalidraw or your app (of course),
the UI should adapt to the browser's support situation.
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+If the Native File System API is supported (`if ('chooseFileSystemEntries in window) {}`)
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:330]:
```suggestion
If the Native File System API is supported (`if ('chooseFileSystemEntries' in window) {}`)
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+you can show a **Save As** button apart from a **Save** button.
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:331]:
```suggestion
you can show a **Save As** button in addition to a **Save** button.
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 20]
```diff
@@ -0,0 +1,366 @@
...
+The screenshots below show the difference of Excalidraw's responsive main app toolbar on iPhone and on Chrome desktop.
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:332]:
```suggestion
The screenshots below show the difference between Excalidraw's responsive main app toolbar on iPhone and on Chrome desktop.
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 21]
```diff
@@ -0,0 +1,366 @@
...
+  <figcaption class="w-figcaption">
+    Excalidraw app toolbar on Chrome  with a <strong>Save</strong> and a focused <strong>Save As</strong> button.
+  </figcaption>
+</figure>
+
+## Conclusions
+
+Working with native files technically works on all modern browsers.
+On browsers that support the Native File System API, you can make the experience better by allowing
+for true saving and over-saving (not just downloading) of files and
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:353]:
```suggestion
for true saving and over writing (not just downloading) of files and
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 22]
```diff
@@ -0,0 +1,366 @@
...
+by letting your users create new files wherever they want;
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:354]:
```suggestion
by letting your users create new files wherever they want,
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 23]
```diff
@@ -0,0 +1,366 @@
...
+all while still remaining functional on browsers that do not support the Native File System API.
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:355]:
```suggestion
all while remaining functional on browsers that do not support the Native File System API.
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 24]
```diff
@@ -0,0 +1,366 @@
...
+The 💾&nbsp;[browser-nativefs](https://github.com/GoogleChromeLabs/browser-nativefs) makes your life easier
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:356]:
```suggestion
The [browser-nativefs](https://github.com/GoogleChromeLabs/browser-nativefs) makes your life easier
```

**Comment by jpmedley** [comment]
I almost forgot. I understand why you lead with legacy APIs. I think it would be better to put the Native File System stuff first. Always lead with the newest.

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 25]
```diff
@@ -0,0 +1,366 @@
...
+element.
+In its simplest form, opening a file can look something like in the code sample below.
+You get back a [`FileList`](https://developer.mozilla.org/en-US/docs/Web/API/FileList),
+which in the case below consists of just one
+[`File`](https://developer.mozilla.org/en-US/docs/Web/API/File).
+A `File` is a specific kind of [`Blob`](https://developer.mozilla.org/en-US/docs/Web/API/Blob),
+and can be used in any context that a Blob can.
+
+```js
+const openFile = async () => {
```
[the remark this answers is not in the document]
Comment (tomayac) [reply src/site/content/en/blog/browser-nativefs/index.md:44]: In this case it's what the function does, it opens a file. `const openedFile = await openFile()`.

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 26]
```diff
@@ -0,0 +1,366 @@
...
+
+### Saving files
+
+Saving files is a similarly straight-forward operation.
+From a file handle, you create a writable stream via `createWritable()`,
+then you write the Blob data by calling the stream's `write()` method,
+and finally you need to close the stream by calling its `close()` method.
+
+```js
+const saveFile = async (blob) => {
```
[the remark this answers is not in the document]
Comment (tomayac) [reply src/site/content/en/blog/browser-nativefs/index.md:138]: Good catch. I make it return the handle.

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 27]
```diff
@@ -0,0 +1,366 @@
...
+  try {
+    const handle = await window.chooseFileSystemEntries({
+      type: 'save-file',
+    });
+    const writable = await handle.createWritable();
+    await writable.write(blob);
+    await writable.close();
```
Comment (tomayac) [inline src/site/content/en/blog/browser-nativefs/index.md:145]:
```suggestion
    await writable.close();
    return handle;
```

File: src/site/content/en/blog/browser-nativefs/index.md (hunk 16 above)
Comment (tomayac) [reply src/site/content/en/blog/browser-nativefs/index.md:295]: Again good catch.

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 28]
```diff
@@ -0,0 +1,366 @@
...
+---
+
+Browsers have been able to deal with files and directories for a long time.
+The [File API](https://w3c.github.io/FileAPI/)
+provides features for representing file objects in web applications,
+as well as programmatically selecting them and accessing their data.
+The moment you look closer, though, all that glitters is not gold.
+
+## The traditional way of dealing with files
+
```
Comment (tomayac) [inline src/site/content/en/blog/browser-nativefs/index.md:30]:
```suggestion

{% Aside %}
  If you know how it used to work the old way, you can
  [jump down straight to the new way](#the-native-file-system-api).
{% endAside %}

```

**Comment by tomayac** [comment]
> I almost forgot. I understand why you lead with legacy APIs. I think it would be better to put the Native File System stuff first. Always lead with the newest.

I've added a "jump" link so folks aware of how it works can quickly jump down.

**Comment by tomayac** [comment]
Thanks for the review, Joe, I have addressed all your remarks. PTAL.

**Comment by tomayac** [comment]
> @tomayac I approved this prematurely. Web.dev are getting strict about image requirements. Can you please make the image changes below.
> 
> (This message was auto-generated by the web.dev content review extension)
> 
> * [ ]  Resize the hero image: https://web.dev/handbook/markup-media/#hero
> * [ ]  Resize the width of the following images to 1600px or less:
>   
>   * [ ]  caniuse.png
>   * [ ]  chrome-modify.png
>   * [ ]  chrome-oversave.png
>   * [ ]  chrome-save-as.png

Again an issue in the review extension, all files have the correct dimensions.

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 29]
```diff
@@ -0,0 +1,371 @@
...
+```js
+const openFile = async () => {
+  try {
+    const handle = await window.chooseFileSystemEntries();
+    return handle.getFile();
+  } catch (err) {
+    console.error(err.name, err.message);
+  }
+};
```
Comment (tomayac) [inline src/site/content/en/blog/browser-nativefs/index.md:127]:
```suggestion
opening a file is a matter of one call to the `window.showOpenFilePicker()` method.
This call returns a file handle, from which you can get the actual `File` via the `getFile()` method.

```js
const openFile = async () => {
  try {
    // Always returns an array.
    const [handle] = await window. showOpenFilePicker();
    return handle.getFile();
  } catch (err) {
    console.error(err.name, err.message);
  }
};
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 30]
```diff
@@ -0,0 +1,371 @@
...
+```
+
+### Opening directories
+
+Open a directory by passing an options object like `{type: 'open-directory'}` to
+`chooseFileSystemEntries()` that makes directories selectable in the file dialog box.
```
Comment (tomayac) [inline src/site/content/en/blog/browser-nativefs/index.md:133]:
```suggestion
Open a directory by calling
`window.showDirectoryPicker()` that makes directories selectable in the file dialog box.
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 31]
```diff
@@ -0,0 +1,371 @@
...
+From a file handle, you create a writable stream via `createWritable()`,
+then you write the Blob data by calling the stream's `write()` method,
+and finally you close the stream by calling its `close()` method.
+
+```js
+const saveFile = async (blob) => {
+  try {
+    const handle = await window.chooseFileSystemEntries({
+      type: 'save-file',
+    });
```
Comment (tomayac) [inline src/site/content/en/blog/browser-nativefs/index.md:147]:
```suggestion
    const handle = await window.showSaveFilePicker({
      types: [{
        accept: {
          // Omitted
        },
      }],
    });
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 32]
```diff
@@ -0,0 +1,376 @@
...
+  });
+  return loadFromBlob(blob);
+};
+```
+
+### UI considerations
+
+Whether in Excalidraw or your app,
+the UI should adapt to the browser's support situation.
+If the Native File System API is supported (`if ('chooseFileSystemEntries' in window) {}`)
```
Comment (tomayac) [inline src/site/content/en/blog/browser-nativefs/index.md:340]:
```suggestion
If the Native File System API is supported (`if ('showOpenFilePicker' in window) {}`)
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 33]
```diff
@@ -0,0 +1,376 @@
...
+  - thomassteiner
+description: |
+  All modern browsers can read native files and directories; however,
+  true write access, that is, more than just downloading files,
+  is limited to browsers that implement the Native File System API.
+  This post introduces a support library called browser-nativefs
+  that acts as an abstraction layer on top of the Native File System API
+  and that transparently falls back to legacy approaches for dealing with files.
+scheduled: true
+date: 2020-06-30
```
Comment (tomayac) [inline src/site/content/en/blog/browser-nativefs/index.md:14]:
```suggestion
date: 2020-07-25
```

File: src/site/content/en/blog/browser-nativefs/index.md [hunk 34]
```diff
@@ -0,0 +1,376 @@
...
+  - thomassteiner
+description: |
+  All modern browsers can read native files and directories; however,
+  true write access, that is, more than just downloading files,
+  is limited to browsers that implement the Native File System API.
+  This post introduces a support library called browser-nativefs
+  that acts as an abstraction layer on top of the Native File System API
+  and that transparently falls back to legacy approaches for dealing with files.
+scheduled: true
+date: 2020-07-25
```
Comment (jpmedley) [inline src/site/content/en/blog/browser-nativefs/index.md:14]:
```suggestion
date: 2020-07-27
```

**Pull request merged**